src/consensus/raft/raft.h

```diff
@@ -579,8 +563,13 @@ namespace raft
         return;
       }
 
+      const auto prev_term = get_term_internal(r.prev_idx);
```

Comment:
Interesting one here. It seems that this PR has uncovered a behaviour that was hidden by the fact that we never sent the first few append entries message.

On late join (this appeared on the full test suite after a recovery), it is possible that the primary sends several AE messages with the very last index `V` in its log. The late joiner will reply with a `false` AER message as it does not have the latest message `V - 1` in its log, to which the primary will respond with many AE messages, from 0 to `V`, which will be happily deserialised by the late joiner. *All is good so far.* 

However, it is possible that one of the `false` AER was in flight and only read by the primary after it has sent the AEs from 0 to `V`. In that case, the primary will send all these entries again, which is wasteful (but that's not the main issue here, and snapshots should somehow address that). 

All the new AEs will not be deserialised by the late joiner (it has already deserialised those entries) and based on the code on the left-hand side of this diff, the election timeout will not be restarted (it is guarded by the `if (r.prev_idx < commit_idx)` check on line `545`), and which will eventually cause an election. 

I've shuffled the orders of the checks around here so that the election timeout is re-started even if the AE index is behind our commit index. However, if the term is stale, we don't want to reset our election timeout. This seems to be what [the Raft reference implementation does too](https://github.com/logcabin/logcabin/blob/master/Server/RaftConsensus.cc#L1307).

Reply:
That’s a very good catch!